lib/mongo/collection.ex

```diff
@@ -580,17 +580,23 @@ defmodule Mongo.Collection do
               unquote(attribute_names),
               %__MODULE__{},
               fn name, result ->
-                Map.put(result, name, map[Atom.to_string(name)])
+                Map.put(result, name, map[Atom.to_string(@attributes[name][:name])])
               end
             )
 
           struct =
             unquote(embed_ones)
-            |> Enum.map(fn {name, mod} -> {name, mod.load(map[Atom.to_string(name)])} end)
+            |> Enum.map(fn {name, mod} ->
+              {_, _, opts} = List.keyfind(@embed_ones, name, 0)
+              {name, mod.load(map[Atom.to_string(opts[:name])])}
+            end)
```

Comment:
Same here. I would not use a search function to get the options for a specific `embed_ones`. If we have a lot of embed_ones then this smells a little bit like O(n^2) runtime behavior. We should move the work to find the attribute name outside of the function, so we use it like this:

```elixir
|> Enum.map(fn {name, mod, src_name} -> {name, mod.load(map[src_name])} end)
```